File: Data Structures & Algorithms/count-number-of-islands/submission-3.cpp
class Solution {
public:
    bool isValid(int i, int j, int m, int n) {
        if (i < 0 || i >= m) return false;
        if (j < 0 || j >= n) return false;
        return true;
    }

    void dfs(int i, int j, vector<vector<bool>>& vis, vector<vector<char>>& grid, int m, int n) {
        if (!isValid(i, j, m, n)) return;
        if (vis[i][j]) return;
        if (grid[i][j] == '0') return;

        vis[i][j] = true;

        dfs(i+1, j, vis, grid, m, n);
        dfs(i-1, j, vis, grid, m, n);
        dfs(i, j+1, vis, grid, m, n);
        dfs(i, j-1, vis, grid, m, n);
    }

    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();

        int count = 0;
        vector<vector<bool>> vis(m, vector<bool>(n, false));

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!vis[i][j] && grid[i][j] == '1') {
                    count++;
                    dfs(i, j, vis, grid, m, n);
                }
            }
        }
        return count;
    }
};

File: Data Structures & Algorithms/rotting-fruit/submission-3.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:

        delRow = [-1, 0, 1, 0]
        delCol = [0,-1,0,1]

        fresh = 0 
        m = len(grid)
        n = len(grid[0])

        minutes = 0 

        q = deque()

        def isValid(i,j,m,n):

            if i<0 or i>=m:
                return False
            if j<0 or j>=n:
                return False 
            return True
            

        for i in range(0,m):
            for j in range(0, n):
                if grid[i][j] == 2:
                    q.append((i,j))
                elif grid[i][j] == 1:
                    fresh+=1

        
        while q:

            k = len(q) 
            for _ in range(k):
                row, col = q.popleft()

                for i in range(0, 4):
                    nR, nC = row + delRow[i], col + delCol[i]

                    if isValid(nR, nC, m, n) and grid[nR][nC] == 1:
                        grid[nR][nC] = 2
                        fresh -= 1
                        q.append((nR, nC))
                
            if q:
                minutes+=1
        
        if fresh == 0:
            return minutes 
        else:
            return -1 



                    

               
                            

                    
                 

            

        
        